File: MAHAndroidUpdater/mah-android-updater/src/main/java/com/mobapphome/mahandroidupdater/tools/MAHUpdaterController.java
package com.mobapphome.mahandroidupdater.tools;

import android.content.Context;
import android.content.SharedPreferences;
import android.graphics.Typeface;
import android.support.annotation.NonNull;
import android.support.v4.app.DialogFragment;
import android.support.v4.app.Fragment;
import android.support.v4.app.FragmentActivity;
import android.support.v4.app.FragmentManager;
import android.support.v4.app.FragmentTransaction;
import android.util.Log;
import android.widget.TextView;

import com.mobapphome.mahandroidupdater.MAHRestricterDlg;
import com.mobapphome.mahandroidupdater.MAHUpdaterDlg;
import com.mobapphome.mahandroidupdater.R;
import com.mobapphome.mahandroidupdater.types.ProgramInfo;
import com.mobapphome.mahandroidupdater.types.DlgModeEnum;

public class MAHUpdaterController {
    public static String urlService;
    private static SharedPreferences sharedPref;
    private static String fontName = null;
    private static FragmentActivity act;
    private static boolean initCalled = false;

    private static boolean btnInfoVisibility;
    private static String btnInfoMenuItemTitle;
    private static String btnInfoActionURL;

    /**
     * Initializes MAHAndroidUpdater library
     * @param act Activity which init calls
     * @param urlService Url for services which data about update has placed.
     */
    public static void init(final FragmentActivity act,
                            String urlService) throws NullPointerException {
        MAHUpdaterController.init(act, urlService,true);
    }

    /**
     * Initializes MAHAndroidUpdater library
     * @param act Activity which init calls
     * @param urlService Url for services which data about update has placed.
     * @param btnInfoVisibility If true shows info button
     */
    public static void init(final FragmentActivity act,
                            String urlService,
                            boolean btnInfoVisibility) throws NullPointerException {
        MAHUpdaterController.init(
                act,
                urlService,
                btnInfoVisibility,
                act.getString(R.string.mah_android_upd_info_popup_text),
                Constants.MAH_UPD_GITHUB_LINK);
    }

    /**
     * Initializes MAHAndroidUpdater library
     * @param act Activity which init calls
     * @param urlService Url for services which data about update has placed.
     * @param btnInfoVisibility If true shows info button
     * @param btnInfoMenuItemTitle Title of menu item for info button
     * @param btnInfoActionURL Url to open when clicking to info button or info menu item
     */
    public static void init(
            final FragmentActivity act,
            String urlService,
            boolean btnInfoVisibility,
            String btnInfoMenuItemTitle,
            @NonNull String btnInfoActionURL) throws NullPointerException {
        if (initCalled) {
            return;
        }

        MAHUpdaterController.btnInfoVisibility = btnInfoVisibility;
        MAHUpdaterController.btnInfoMenuItemTitle = btnInfoMenuItemTitle;
        MAHUpdaterController.btnInfoActionURL = btnInfoActionURL;

        MAHUpdaterController.urlService = urlService;
        MAHUpdaterController.act = act;
        if (urlService == null) {
            throw new NullPointerException("urlService not set call init(final Activity act, String urlService) constructor");
        }

        sharedPref = act.getPreferences(Context.MODE_PRIVATE);

        Updater updater = new Updater();
        updater.setUpdaterListiner(new UpdaterListener() {

            @Override
            public void onResponse(ProgramInfo programInfo, final String errorStr) {
                if (programInfo == null) {
                    Log.i(Constants.MAH_ANDROID_UPDATER_LOG_TAG, "MAhUpdater Program info is null");
                    return;
                } else if (!programInfo.isRunMode()) {
                    Log.i(Constants.MAH_ANDROID_UPDATER_LOG_TAG, "MAHUpdter run mode is false");
                    return;
                } else if (programInfo.getUriCurrent() == null) {
                    Log.i(Constants.MAH_ANDROID_UPDATER_LOG_TAG, "MAHUpdter uri_current is null in service. check service");
                    return;
                } else if (programInfo.getVersionCodeCurrent() == -1) {
                    Log.i(Constants.MAH_ANDROID_UPDATER_LOG_TAG, "MAHUpdter version_code_current is -1 in service. check service");
                    return;
                } else {
					Log.i(Constants.MAH_ANDROID_UPDATER_LOG_TAG, "Uri current from service = " + programInfo.getUriCurrent() + "  Uri from package" + act.getApplicationContext().getPackageName());
					Log.i(Constants.MAH_ANDROID_UPDATER_LOG_TAG, "Version from service = " + programInfo.getVersionCodeCurrent() + "  Version from package" + Utils.getVersionCode(act));

                    boolean isRestrictedDlg = false;
                    if (Utils.getVersionCode(act) < programInfo.getVersionCodeMin()) {
                        isRestrictedDlg = true;
                    }
                    if (!programInfo.getUriCurrent().equals(act.getApplicationContext().getPackageName())) {
                        if (Utils.checkPackageIfExists(act, programInfo.getUriCurrent())) {
                            showRestricterDlg(act, DlgModeEnum.OPEN_NEW, programInfo);
                        } else if (isRestrictedDlg) {
                            showRestricterDlg(act, DlgModeEnum.INSTALL, programInfo);
                        } else {
                            showUpdaterDlg(act, DlgModeEnum.INSTALL, programInfo);
                        }
                    } else if (Utils.getVersionCode(act) < programInfo.getVersionCodeCurrent()) {
                        if (isRestrictedDlg) {
                            showRestricterDlg(act, DlgModeEnum.UPDATE, programInfo);
                        } else {
                            showUpdaterDlg(act, DlgModeEnum.UPDATE, programInfo);
                        }
                    } else {
                        Log.i(Constants.MAH_ANDROID_UPDATER_LOG_TAG, "MAHUpdater: There are not any update in service");
                    }
                }
            }
        });
        updater.updateProgramList(act);
        initCalled = true;
    }

    public static void callUpdate() throws NullPointerException {
        if (MAHUpdaterController.urlService == null) {
            throw new NullPointerException("urlService not set call init(final Activity act, String urlService) constructor");
        }


        Updater updater = new Updater();
        updater.setUpdaterListiner(new UpdaterListener() {

            @Override
            public void onResponse(ProgramInfo programInfo, final String errorStr) {

            }
        });
        updater.updateProgramList(act);
    }


    public static void end() {
        Log.i(Constants.MAH_ANDROID_UPDATER_LOG_TAG, "MAHUpdater end called");
        callUpdate();
        initCalled = false;
    }

    static public void testUpdaterDlg(final FragmentActivity act){
        final ProgramInfo prInfoTest = new ProgramInfo();
        prInfoTest.setUpdateInfo("Update info test mode.");
        showUpdaterDlg(act, DlgModeEnum.TEST, prInfoTest);
    }

    static public void testRestricterDlg(final FragmentActivity act){
        final ProgramInfo prInfoTest = new ProgramInfo();
        prInfoTest.setUpdateInfo("Update info test mode. ");
        showRestricterDlg(act, DlgModeEnum.TEST, prInfoTest);
    }

    static private void showUpdaterDlg(final FragmentActivity act, DlgModeEnum mode, final ProgramInfo programInfo) {
        showDlg(act,
                MAHUpdaterDlg.newInstance(programInfo, mode, btnInfoVisibility, btnInfoMenuItemTitle, btnInfoActionURL),
                "fragment_android_updater_dlg");
    }

    static private void showRestricterDlg(final FragmentActivity act, DlgModeEnum mode, final ProgramInfo programInfo) {
        showDlg(act,
                MAHRestricterDlg.newInstance(programInfo, mode, btnInfoVisibility, btnInfoMenuItemTitle, btnInfoActionURL),
                "fragment_restricter_dlg");
    }

    static private void showDlg(FragmentActivity activity, Fragment frag, String fragTag) {

        if (!activity.isFinishing()) {
            FragmentManager fragmentManager = activity.getSupportFragmentManager();
            Fragment fr = fragmentManager.findFragmentByTag(fragTag);
            if (fr != null && !fr.isHidden()) {
                Log.i(Constants.MAH_ANDROID_UPDATER_LOG_TAG, "showDlg  dismissed");
                ((DialogFragment) fr).dismissAllowingStateLoss();
            }

            FragmentTransaction transaction = fragmentManager.beginTransaction();
            transaction.add(frag, fragTag);
            transaction.commitAllowingStateLoss();
        }
    }

    protected static SharedPreferences getSharedPref() {
        return sharedPref;
    }

    public static void setFontTextView(TextView tv) {
        if (fontName == null) {
            return;
        }
        try {
            Typeface font = Typeface.createFromAsset(tv.getContext().getAssets(), fontName);
            tv.setTypeface(font);
        } catch (RuntimeException r) {
            Log.e("test", "Error " + r.getMessage());
        }
    }

    public static String getFontName() {
        return fontName;
    }

    public static void setFontName(String fontName) {
        MAHUpdaterController.fontName = fontName;
    }
}

File: MAHAndroidUpdater/mah-android-updater/src/main/java/com/mobapphome/mahandroidupdater/tools/MAHUpdaterController.kt
package com.mobapphome.mahandroidupdater.tools

import android.content.Context
import android.content.SharedPreferences
import android.graphics.Typeface
import android.support.v4.app.DialogFragment
import android.support.v4.app.Fragment
import android.support.v4.app.FragmentActivity
import android.util.Log
import android.widget.TextView

import com.mobapphome.mahandroidupdater.MAHRestricterDlg
import com.mobapphome.mahandroidupdater.MAHUpdaterDlg
import com.mobapphome.mahandroidupdater.R

object MAHUpdaterController {
    var urlService: String? = null
    var updateInfoResolver: IUpdateInfoResolver? = null
    var sharedPref: SharedPreferences? = null
    var fontName: String? = null
    private var act: FragmentActivity? = null
    private var initCalled = false

    private var btnInfoVisibility: Boolean = false
    private var btnInfoMenuItemTitle: String? = null
    private var btnInfoActionURL: String? = null

    /**
     * Initializes MAHAndroidUpdater library
     * @param act Activity which init calls
     * *
     * @param urlService Url for services which data about update has placed.
     * *
     * @param updateInfoResolver Object wich implementing it you can get data from your own structed web service.
     * *
     * @param btnInfoVisibility If true shows info button
     * *
     * @param btnInfoMenuItemTitle Title of menu item for info button
     * *
     * @param btnInfoActionURL Url to open when clicking to info button or info menu item
     */
    @Throws(NullPointerException::class)
    fun init(
            act: FragmentActivity,
            urlService: String?,
            updateInfoResolver: IUpdateInfoResolver? = null,
            btnInfoVisibility: Boolean = true,
            btnInfoMenuItemTitle: String = act.getString(R.string.mah_android_upd_info_popup_text),
            btnInfoActionURL: String = Constants.MAH_UPD_GITHUB_LINK) {
        if (initCalled) {
            return
        }

        MAHUpdaterController.btnInfoVisibility = btnInfoVisibility
        MAHUpdaterController.btnInfoMenuItemTitle = btnInfoMenuItemTitle
        MAHUpdaterController.btnInfoActionURL = btnInfoActionURL

        MAHUpdaterController.urlService = urlService
        MAHUpdaterController.updateInfoResolver = updateInfoResolver
        MAHUpdaterController.act = act

        if (urlService == null && updateInfoResolver == null) {
            throw NullPointerException("At least one of these urlService or updateInfoResolver variables " + "\n has to be set in init(final Activity act, String urlService, IUpdateInfoResolver updateInfoResolver) method")
        }

        if (urlService != null && updateInfoResolver != null) {
            throw RuntimeException("Can't use urlService and updateInfoResolver at the same time, choose one")
        }

        sharedPref = act.getPreferences(Context.MODE_PRIVATE)

        val updater = Updater()
        updater.updaterListiner = object : UpdaterListener {

            override fun onResponse(programInfo: ProgramInfo, errorStr: String) {
                when {
                    programInfo == null -> Log.i(Constants.MAH_ANDROID_UPDATER_LOG_TAG, "MAhUpdater Program info is null")
                    !programInfo.isRunMode -> Log.i(Constants.MAH_ANDROID_UPDATER_LOG_TAG, "MAHUpdter run mode is false")
                    programInfo.uriCurrent == null -> Log.i(Constants.MAH_ANDROID_UPDATER_LOG_TAG, "MAHUpdter uri_current is null in service. check service")
                    programInfo.versionCodeCurrent == -1 -> Log.i(Constants.MAH_ANDROID_UPDATER_LOG_TAG, "MAHUpdter version_code_current is -1 in service. check service")
                    else -> {
                        Log.i(Constants.MAH_ANDROID_UPDATER_LOG_TAG, "Uri current from service = " + programInfo.uriCurrent + "  Uri from package" + act.applicationContext.packageName)
                        Log.i(Constants.MAH_ANDROID_UPDATER_LOG_TAG, "Version from service = " + programInfo.versionCodeCurrent + "  Version from package" + getVersionCode(act))

                        val isRestrictedDlg = if (getVersionCode(act) < programInfo.versionCodeMin) true else false

                        when {
                            programInfo.uriCurrent != act.applicationContext.packageName ->
                                when {
                                    checkPackageIfExists(act, programInfo.uriCurrent) -> showRestricterDlg(act, DlgModeEnum.OPEN_NEW, programInfo)
                                    isRestrictedDlg -> showRestricterDlg(act, DlgModeEnum.INSTALL, programInfo)
                                    else -> showUpdaterDlg(act, DlgModeEnum.INSTALL, programInfo)
                                }
                            getVersionCode(act) < programInfo.versionCodeCurrent ->
                                when {
                                    isRestrictedDlg -> showRestricterDlg(act, DlgModeEnum.UPDATE, programInfo)
                                    else -> showUpdaterDlg(act, DlgModeEnum.UPDATE, programInfo)
                                }
                            else -> Log.i(Constants.MAH_ANDROID_UPDATER_LOG_TAG, "MAHUpdater: There are not any update in service")
                        }
                    }
                }
            }
        }

        updater.updateProgramList(act)
        initCalled = true
    }

    @Throws(NullPointerException::class)
    fun callUpdate() {
        if (MAHUpdaterController.urlService == null && MAHUpdaterController.updateInfoResolver == null) {
            return
            //throw new NullPointerException("urlService not set call init(final Activity act, String urlService) constructor");
        }

        val updater = Updater()
        updater.updaterListiner = object : UpdaterListener {

            override fun onResponse(programInfo: ProgramInfo, errorStr: String) {

            }
        }

        updater.updateProgramList(act!!)
    }


    fun end() {
        Log.i(Constants.MAH_ANDROID_UPDATER_LOG_TAG, "MAHUpdater end called")
        callUpdate()
        initCalled = false
    }

    fun testUpdaterDlg(act: FragmentActivity) {
        val prInfoTest = ProgramInfo(updateInfo = "Update info test mode.")
        showUpdaterDlg(act, DlgModeEnum.TEST, prInfoTest)
    }

    fun testRestricterDlg(act: FragmentActivity) {
        val prInfoTest = ProgramInfo(updateInfo = "Update info test mode. ")
        showRestricterDlg(act, DlgModeEnum.TEST, prInfoTest)
    }

    private fun showUpdaterDlg(act: FragmentActivity, mode: DlgModeEnum, programInfo: ProgramInfo) {
        showDlg(act,
                MAHUpdaterDlg.newInstance(programInfo, mode, btnInfoVisibility, btnInfoMenuItemTitle, btnInfoActionURL),
                "fragment_android_updater_dlg")
    }

    private fun showRestricterDlg(act: FragmentActivity, mode: DlgModeEnum, programInfo: ProgramInfo) {
        showDlg(act,
                MAHRestricterDlg.newInstance(programInfo, mode, btnInfoVisibility, btnInfoMenuItemTitle, btnInfoActionURL),
                "fragment_restricter_dlg")
    }

    private fun showDlg(activity: FragmentActivity, frag: Fragment, fragTag: String) {

        if (!activity.isFinishing) {
            val fragmentManager = activity.supportFragmentManager
            val fr = fragmentManager.findFragmentByTag(fragTag)
            if (fr != null && !fr.isHidden) {
                Log.i(Constants.MAH_ANDROID_UPDATER_LOG_TAG, "showDlg  dismissed")
                (fr as DialogFragment).dismissAllowingStateLoss()
            }

            val transaction = fragmentManager.beginTransaction()
            transaction.add(frag, fragTag)
            transaction.commitAllowingStateLoss()
        }
    }
}
